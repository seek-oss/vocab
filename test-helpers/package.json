{
  "name": "@vocab-private/test-helpers",
  "private": true,
  "version": "1.0.0",
  "author": "SEEK",
  "license": "MIT",
  "main": "src/index.ts",
  "devDependencies": {
    "@fixtures/direct": "*",
    "@fixtures/phrase": "*",
    "@fixtures/server": "*",
    "@fixtures/simple": "*",
    "@fixtures/translation-types": "*",
    "@types/wait-on": "^5.3.1",
    "@types/webpack-dev-server": "^3.11.6",
    "@vocab/core": "*",
    "wait-on": "^5.2.0",
    "webpack": "^5.11.0",
    "webpack-dev-server": "^4.11.1"
  }
}
