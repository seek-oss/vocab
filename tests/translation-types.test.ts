import { compileFixtureTranslations } from '@vocab-private/test-helpers';
import { promises } from 'fs';

describe('Translation types', () => {
  describe('all message types', () => {
    it('should generate types for translations', async () => {
      await compileFixtureTranslations('translation-types');
      const compiledTranslations = require.resolve(
        '@vocab-fixtures/translation-types/src/all-message-types/.vocab/index.ts',
      );

      expect(
        await promises.readFile(compiledTranslations, 'utf8'),
      ).toMatchSnapshot();
    });
  });

  describe('a single message', () => {
    it('should generate types for translations with no type import from @vocab/core', async () => {
      await compileFixtureTranslations('translation-types');
      const compiledTranslations = require.resolve(
        '@vocab-fixtures/translation-types/src/no-vocab-types-import/.vocab/index.ts',
      );

      expect(
        await promises.readFile(compiledTranslations, 'utf8'),
      ).toMatchSnapshot();
    });
  });
});
