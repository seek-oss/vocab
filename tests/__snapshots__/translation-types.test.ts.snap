// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Translation types a single message should generate types for translations with no type import from @vocab/core 1`] = `
"// This file is automatically generated by Vocab.
// To make changes update translation.json files directly.

import { createLanguage, createTranslationFile } from '@vocab/core/runtime';

const translations = createTranslationFile<'en', { Simple: () => 'Simple' }>({
  en: createLanguage({ Simple: 'Simple' }),
});

export default translations;
"
`;

exports[`Translation types all message types should generate types for translations 1`] = `
"// This file is automatically generated by Vocab.
// To make changes update translation.json files directly.

import type { FormatXMLElementFn, StringWithSuggestions } from '@vocab/core';
import { createLanguage, createTranslationFile } from '@vocab/core/runtime';

const translations = createTranslationFile<
  'en',
  {
    Simple: () => 'Simple';
    Param: (values: { candidateName: string }) => string;
    Tag: <T = string>(values: {
      Bold: FormatXMLElementFn<T>;
    }) => string | T | Array<string | T>;
    Number: (values: { numCats: number }) => string;
    Date: (values: { start: Date | number }) => string;
    Time: (values: { expires: Date | number }) => string;
    'Nested tags': <T = string>(values: {
      Bold: FormatXMLElementFn<T>;
      Italic: FormatXMLElementFn<T>;
    }) => string | T | Array<string | T>;
    Plural: (values: { itemCount: number }) => string;
    'Plural with nested params': (values: {
      numDogs: number;
      singularPrefix: string;
      pluralPrefix: string;
    }) => string;
    'Plural with self-referential param': (values: { foo: number }) => string;
    'Select ordinal': (values: { year: number }) => string;
    'Select ordinal with self-referential param': (values: {
      year: number;
    }) => string;
    Select: (values: { foo: StringWithSuggestions<'a' | 'b'> }) => string;
    'Select with nested params': (values: {
      foo: StringWithSuggestions<'a' | 'b'>;
      bar: string;
      baz: string;
    }) => string;
    'Select with pound element': (values: {
      foo: StringWithSuggestions<'a' | 'b'>;
    }) => string;
    'Select with self-referential param': (values: {
      foo: StringWithSuggestions<'a' | 'b'>;
    }) => string;
  }
>({
  en: createLanguage({
    Simple: 'Simple',
    Param: 'Welcome {candidateName}!',
    Tag: 'Foo bar <Bold>baz</Bold>',
    Number: 'I have {numCats, number} cats',
    Date: 'Sale begins {start, date, medium}',
    Time: 'Coupon expires at {expires, time, short}',
    'Nested tags': 'The quick <Bold>brown <Italic>fox</Italic></Bold>',
    Plural: '{itemCount, plural, one {# item} other {# items}}',
    'Plural with nested params':
      '{numDogs, plural, one {# {singularPrefix} dog} other {# {pluralPrefix} dogs}}',
    'Plural with self-referential param':
      '{foo, plural, zero {a} one {b} other {{foo} c}}',
    'Select ordinal':
      "It's my cat's {year, selectordinal, one {#st} two {#nd} few {#rd} other {#th}} birthday!",
    'Select ordinal with self-referential param':
      "It's my cat's {year, selectordinal, one {{year}st} two {{year}nd} few {{year}rd} other {{year}th}} birthday!",
    Select: '{foo, select, a {a} b {b} other {c}}',
    'Select with nested params':
      '{foo, select, a {{bar} a} b {{baz} b} other {c}}',
    'Select with pound element': '{foo, select, a {# a} b {# b} other {c}}',
    'Select with self-referential param':
      '{foo, select, a {a} b {b} other {{foo} c}}',
  }),
});

export default translations;
"
`;
